### PR TITLE
Rbsfxirjtp 148 market service subscription mechanism

### PR DESCRIPTION
**Pull request by nikolanguen**
Added market subscription mechanism.

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -1,46 +0,0 @@
-# This workflow will build a Java project with Maven
```
Comment (trayan-muchev) [inline .github/workflows/deploy.yml:1]: Why have you deleted this file?

File: .github/workflows/maven.yml [hunk 2]
```diff
@@ -1,29 +0,0 @@
-# This workflow will build a Java project with Maven
```
Comment (trayan-muchev) [inline .github/workflows/maven.yml:1]: Why have you deleted this file as well?

File: aurora/src/main/java/com/market/banica/aurora/service/AuroraServiceImpl.java [hunk 3]
```diff
@@ -6,15 +6,15 @@
 import org.springframework.stereotype.Service;
 
 @Service
-public class AuroraService extends AuroraServiceGrpc.AuroraServiceImplBase {
+public class AuroraServiceImpl extends AuroraServiceGrpc.AuroraServiceImplBase {
```
Comment (trayan-muchev) [inline aurora/src/main/java/com/market/banica/aurora/service/AuroraServiceImpl.java:9]: You have compilation errors in this class.

File: market-generator/src/main/java/com/market/banica/generator/service/Subscription.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package com.market.banica.generator.service;
+
+import io.grpc.stub.StreamObserver;
+
+public interface Subscription<T, S> {
```
Comment (trayan-muchev) [inline market-generator/src/main/java/com/market/banica/generator/service/Subscription.java:5]: Rename to SubscriptionManager or MarketSubscriptionManager

File: market-generator/src/main/java/com/market/banica/generator/service/MarketSubscriptionImpl.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Component
+public class MarketSubscriptionImpl implements Subscription<MarketDataRequest, TickResponse> {
```
Comment (trayan-muchev) [inline market-generator/src/main/java/com/market/banica/generator/service/MarketSubscriptionImpl.java:19]: Rename to SubscriptionManager or MarketSubscriptionManager depending on the interface name.

File: market-generator/src/main/java/com/market/banica/generator/service/MarketSubscriptionManager.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    private final Logger LOGGER = LoggerFactory.getLogger(MarketSubscriptionManager.class);
+    private final Map<String, HashSet<StreamObserver<TickResponse>>> subscriptions = new ConcurrentHashMap<>();
+
+    @Override
+    public void subscribe(MarketDataRequest request, StreamObserver<TickResponse> responseObserver) {
+        String goodName = getRequestGoodName(request);
+        if (!StringUtil.isNullOrEmpty(goodName)) {
+            LOGGER.debug("{} Requested for subscription for good: {}.", responseObserver, goodName);
+            subscriptions.compute(goodName, (key, value) -> {
```
Comment (trayan-muchev) [inline market-generator/src/main/java/com/market/banica/generator/service/MarketSubscriptionManager.java:29]: Extract it in a separate method. Suggestion for method name: addSubscriber.

File: market-generator/src/main/java/com/market/banica/generator/service/MarketSubscriptionManager.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            LOGGER.warn("Illegal value {} for good by request: {}.", goodName, request);
+        }
+    }
+
+    @Override
+    public void unsubscribe(MarketDataRequest request, StreamObserver<TickResponse> responseObserver) {
+        String goodName = getRequestGoodName(request);
+        if (!StringUtil.isNullOrEmpty(goodName)) {
+            LOGGER.debug("Request to unsubscribe from {} for good: {}.", responseObserver, goodName);
+            this.subscriptions.compute(goodName, (key, value) -> {
```
Comment (trayan-muchev) [inline market-generator/src/main/java/com/market/banica/generator/service/MarketSubscriptionManager.java:44]: Extract in a separate method. Suggestion for method name: removeSubscriber.

File: market-generator/src/main/java/com/market/banica/generator/service/MarketSubscriptionManager.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            LOGGER.warn("Illegal value {} for good by request: {}.", goodName, request);
+        }
+    }
+
+    @Override
+    public void notifySubscribers(TickResponse response) {
+        String goodName = getTickResponseGoodName(response);
+        if (!StringUtil.isNullOrEmpty(goodName)) {
+            Set<StreamObserver<TickResponse>> subscribers = subscriptions.get(goodName);
+            if (subscribers != null) {
```
Comment (trayan-muchev) [inline market-generator/src/main/java/com/market/banica/generator/service/MarketSubscriptionManager.java:68]: Extract in method. Suggested name sendNotification.

File: market-generator/src/test/java/market/MarketServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        List<TickResponse> ticks = Arrays.asList(TickResponse.newBuilder().setGoodName("firstTick").build(),
+                TickResponse.newBuilder().setGoodName("secondTick").build());
+
+        Mockito.when(marketSubscriptionServiceImpl.getRequestGoodName(MARKET_DATA_REQUEST_BANICA)).thenReturn(GOOD_BANICA);
+        Mockito.when(tickGenerator.generateTicks(GOOD_BANICA)).thenReturn(ticks);
+
+        marketService.subscribeForItem(MARKET_DATA_REQUEST_BANICA, subscriberOne);
+
+        verify(subscriberOne, times(2)).onNext(any());
+        verify(subscriberOne, times(1)).onCompleted();
```
Comment (trayan-muchev) [inline market-generator/src/test/java/market/MarketServiceImplTest.java:53]: Add verifications for marketSubscriptionServiceImpl and tickGenerator.

File: market-generator/src/test/java/market/MarketServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    private TickGenerator tickGenerator;
+
+    private static final String GOOD_BANICA = "Banica";
+
+    private static final MarketDataRequest MARKET_DATA_REQUEST_BANICA = MarketDataRequest.newBuilder().setGoodName(GOOD_BANICA).build();
+
+    private final StreamObserver<TickResponse> subscriberOne = mock(StreamObserver.class);
+
+    @Test
+    void subscribeForItem() {
```
Comment (trayan-muchev) [inline market-generator/src/test/java/market/MarketServiceImplTest.java:43]: Suggestion: Use the following convention for test names: "whatTheTestDoes/WhatParamsItTakes/Result
Example: subscribeForItemWithValidDataInputExcutesWithSuccess

File: market-generator/src/test/java/market/MarketSubscriptionManagerTest.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+
+    private static final String GOOD_EGGS = "Eggs";
+
+    private static final MarketDataRequest MARKET_DATA_REQUEST_BANICA = MarketDataRequest.newBuilder().setGoodName(GOOD_BANICA).build();
+
+    private static final MarketDataRequest MARKET_DATA_REQUEST_INVALID = MarketDataRequest.newBuilder().build();
+
+
+    @Test
+    void subscribeShouldAddNewSubscriberToSubscriptions() {
```
Comment (trayan-muchev) [inline market-generator/src/test/java/market/MarketSubscriptionManagerTest.java:42]: Suggestion: Use the following convention for test names: "whatTheTestDoes/WhatParamsItTakes/Result
Example: subscribeForItemWithValidDataInputExcutesWithSuccess

This comment is valid for all test names.

File: market-generator/pom.xml [hunk 12]
```diff
@@ -31,6 +31,18 @@
...
         </dependency>
 
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.junit.vintage</groupId>
+                    <artifactId>junit-vintage-engine</artifactId>
```
Comment (zaraandreevaa) [inline market-generator/pom.xml:41]: why do you need this exclusion?

File: market-generator/src/main/java/com/market/banica/generator/exception/NoSuchGoodException.java [hunk 13]
```diff
@@ -0,0 +1,7 @@
+package com.market.banica.generator.exception;
+
+public class NoSuchGoodException extends RuntimeException {
```
Comment (zaraandreevaa) [inline market-generator/src/main/java/com/market/banica/generator/exception/NoSuchGoodException.java:3]: I would suggest to create more generic name of the exception to handle more cases - the message could specify that the good is not found

File: market-generator/src/test/java/market/MarketSubscriptionManagerTest.java [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+import io.grpc.stub.StreamObserver;
+import org.junit.Assert;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.util.HashSet;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (zaraandreevaa) [inline market-generator/src/test/java/market/MarketSubscriptionManagerTest.java:16]: there are settings in intellij for not importing with *, it is not a good practice

File: market-generator/src/test/java/market/MarketSubscriptionManagerTest.java [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+@ExtendWith(MockitoExtension.class)
+class MarketSubscriptionManagerTest {
```
Comment (zaraandreevaa) [inline market-generator/src/test/java/market/MarketSubscriptionManagerTest.java:23]: please add Arrange/Act/Assert or given/when/then for each of your tests to be more clear and understandable and follow this for all the tests

File: market-generator/src/test/java/market/MarketSubscriptionManagerTest.java [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+        Assert.assertEquals(expected, actual);
+    }
+
+    @Test
+    void subscribeShouldNotAddNewSubscriberToSubscriptionsIfIllegalGoodName() {
+        marketSubscriptionManager.subscribe(MARKET_DATA_REQUEST_INVALID, subscriberOne);
+        marketSubscriptionManager.subscribe(MARKET_DATA_REQUEST_BANICA, subscriberTwo);
+
+        int actual = marketSubscriptionManager.getSubscribers(MARKET_DATA_REQUEST_BANICA.getGoodName()).size();
+        int expected = 1;
```
Comment (zaraandreevaa) [inline market-generator/src/test/java/market/MarketSubscriptionManagerTest.java:57]: expected should be declared before the actual execution as part of Arrange/given

File: market-generator/src/test/java/market/MarketSubscriptionManagerTest.java [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+
+        Assert.assertEquals(expected, actual);
+    }
+
+    @Test
+    public void unsubscribeShouldRemoveMapWhenOnlyOneSubscriberInside() {
+        marketSubscriptionManager.subscribe(MARKET_DATA_REQUEST_BANICA, subscriberOne);
```
Comment (zaraandreevaa) [inline market-generator/src/test/java/market/MarketSubscriptionManagerTest.java:88]: this test should be refactored as with the current implementation is hard to understand if you are testing subscribe or unsubscribe and should be re-structured as the comments mentioned above

**Review by trayan-muchev — APPROVED** [review]

**Pull request merged**